File: openage/util/fslike/path.py
# Copyright 2015-2017 the openage authors. See copying.md for legal info.

"""
Provides Path, which is analogous to pathlib.Path,
and the type of FSLikeObject.root.
"""

from io import UnsupportedOperation, TextIOWrapper


class Path:
    """
    Implements an interface somewhat similar to that of pathlib.Path,
    but some methods are missing or have different usage, and some new have
    been added.

    Represents a specific path in a given FS-Like object; mostly, that
    object's member methods are simply wrapped.

    fsobj: fs-like object that is e.g. a cab-archive, a real
           Directory("/lol"), or anything that is like some filesystem.

    parts: starting path in the above fsobj,
           e.g. ["folder", "file"],
           or "folder/file"
           or b"folder/file".
    """

    # We're re-implementing pathlib.Path's interface, we have no choice about
    # this. Also, some of these are properties, so the effective count is
    # lower.
    # pylint: disable=too-many-public-methods

    def __init__(self, fsobj, parts=None):
        if isinstance(parts, str):
            parts = parts.encode()

        if isinstance(parts, (bytes, bytearray)):
            parts = parts.split(b'/')

        if parts is None:
            parts = []

        if not isinstance(parts, (list, tuple)):
            raise ValueError("path parts must be str, bytes, list or tuple, "
                             "but not: %s" % type(parts))

        result = []
        for part in parts:
            if isinstance(part, str):
                part = part.encode()

            if part == b'.' or part == b'':
                pass
            elif part == b'..':
                try:
                    result.pop()
                except IndexError:
                    pass
            else:
                result.append(part)

        self.fsobj = fsobj

        # use tuple instead of list to prevent accidential modification
        self.parts = tuple(result)

    def __str__(self):
        return self.fsobj.pretty(self.parts)

    def __repr__(self):
        if not self.parts:
            return repr(self.fsobj) + ".root"

        return "Path({}, {})".format(repr(self.fsobj), repr(self.parts))

    def exists(self):
        """ True if path exists """
        return self.fsobj.exists(self.parts)

    def is_dir(self):
        """ True if path points to dir (or symlink to one) """
        return self.fsobj.is_dir(self.parts)

    def is_file(self):
        """ True if path points to file (or symlink to one) """
        return self.fsobj.is_file(self.parts)

    def writable(self):
        """ True if path is probably writable """
        return self.fsobj.writable(self.parts)

    def list(self):
        """ Yields path names for all members of this dir """
        yield from self.fsobj.list(self.parts)

    def iterdir(self):
        """ Yields path objects for all members of this dir """
        for name in self.fsobj.list(self.parts):
            yield type(self)(self.fsobj, self.parts + (name,))

    def mkdirs(self):
        """ Creates this path (including parents). No-op if path exists. """
        return self.fsobj.mkdirs(self.parts)

    def open(self, mode="r"):
        """ Opens the file at this path; returns a file-like object. """
        if mode == "rb":
            return self.fsobj.open_r(self.parts)
        elif mode == "r":
            return TextIOWrapper(self.fsobj.open_r(self.parts))
        elif mode == "wb":
            return self.fsobj.open_w(self.parts)
        elif mode == "w":
            return TextIOWrapper(self.fsobj.open_w(self.parts))
        else:
            raise UnsupportedOperation("unsupported open mode: " + mode)

    def open_r(self):
        """ open with mode='rb' """
        return self.fsobj.open_r(self.parts)

    def open_w(self):
        """ open with mode='wb' """
        return self.fsobj.open_w(self.parts)

    def _get_native_path(self):
        """
        return the native path (usable by your kernel) of this path,
        or None if the path is not natively usable.

        Don't use this method directly, use the resolve methods below.
        """
        return self.fsobj.get_native_path(self.parts)

    def _resolve_r(self):
        """
        Flatten the path recursively for read access.
        Used to cancel out some wrappers in between.
        """
        return self.fsobj.resolve_r(self.parts)

    def _resolve_w(self):
        """
        Flatten the path recursively for write access.
        Used to cancel out some wrappers in between.
        """
        return self.fsobj.resolve_w(self.parts)

    def resolve_native_path(self, mode="r"):
        """
        Minimize the path and possibly return a native one.
        Returns None if there was no native path.
        """
        if mode == "r":
            return self.resolve_native_path_r()
        elif mode == "w":
            return self.resolve_native_path_w()
        else:
            raise UnsupportedOperation("unsupported resolve mode: " + mode)

    def resolve_native_path_r(self):
        """
        Resolve the path for read access and possibly return
        a native equivalent.
        If no native path was found, return None.
        """
        resolved_path = self._resolve_r()
        if resolved_path:
            # pylint: disable=protected-access
            return resolved_path._get_native_path()
        return None

    def resolve_native_path_w(self):
        """
        Resolve the path for write access and try to return
        a native equivalent.
        If no native path could be determined, return None.
        """
        resolved_path = self._resolve_w()
        if resolved_path:
            # pylint: disable=protected-access
            return resolved_path._get_native_path()
        return None

    def rename(self, targetpath):
        """ renames to targetpath """
        if self.fsobj != targetpath.fsobj:
            raise UnsupportedOperation("can't rename across two FSLikeObjects")
        return self.fsobj.rename(self.parts, targetpath.parts)

    def rmdir(self):
        """ Removes the empty directory at this path. """
        return self.fsobj.rmdir(self.parts)

    def touch(self):
        """ Creates the file at this path, or updates the timestamp. """
        return self.fsobj.touch(self.parts)

    def unlink(self):
        """ Removes the file at this path. """
        return self.fsobj.unlink(self.parts)

    def removerecursive(self):
        """ Recursively deletes this file or directory. """
        if self.is_dir():
            for path in self.iterdir():
                path.removerecursive()
            self.rmdir()
        else:
            self.unlink()

    @property
    def mtime(self):
        """ Returns the time of last modification of the file or directory. """
        return self.fsobj.mtime(self.parts)

    @property
    def filesize(self):
        """ Returns the file size. """
        return self.fsobj.filesize(self.parts)

    def watch(self, callback):
        """
        Installs 'callback' as callback that gets invoked whenever the file at
        this path changes.

        Returns True if the callback was installed, and false if not
        (e.g. because the some OS limit was reached, or the underlying
         FSLikeObject doesn't support watches).
        """
        return self.fsobj.watch(self.parts, callback)

    def poll_fs_watches(self):
        """ Polls the installed watches for the entire file-system. """
        self.fsobj.poll_watches()

    @property
    def parent(self):
        """ Parent path object. The parent of root is root. """
        return type(self)(self.fsobj, self.parts[:-1])

    @property
    def name(self):
        """ The name of the topmost component (str). """
        return self.parts[-1].decode()

    @property
    def suffix(self):
        """ The last suffix of the name of the topmost component (str). """
        name = self.name
        pos = name.rfind('.')
        if pos <= 0:
            return ""
        return name[pos:]

    @property
    def suffixes(self):
        """ The suffixes of the name of the topmost component (str list). """
        name = self.name
        if name.startswith('.'):
            name = name[1:]
        return ['.' + suffix for suffix in name.split('.')[1:]]

    @property
    def stem(self):
        """ Name without suffix (such that stem + suffix == name). """
        name = self.name
        pos = name.rfind('.')
        if pos <= 0:
            return name

        return name[:pos]

    def joinpath(self, subpath):
        """ Returns path for the given subpath. """
        if isinstance(subpath, str):
            subpath = subpath.encode()

        if isinstance(subpath, bytes):
            subpath = subpath.split(b'/')

        return type(self)(self.fsobj, self.parts + tuple(subpath))

    def __getitem__(self, subpath):
        """ Like joinpath. """
        return self.joinpath(subpath)

    def __truediv__(self, subpath):
        """ Like joinpath. """
        return self.joinpath(subpath)

    def __eq__(self, other):
        """ comparison by fslike and parts """
        return (self.fsobj == other.fsobj) and (self.parts == other.parts)

    def with_name(self, name):
        """ Returns path for differing name (same parent). """
        return self.parent.joinpath(name)

    def with_suffix(self, suffix):
        """ Returns path for different suffix (same parent and stem). """
        if isinstance(suffix, bytes):
            suffix = suffix.decode()

        return self.parent.joinpath(self.stem + suffix)

File: libopenage/util/path.cpp
// Copyright 2015-2017 the openage authors. See copying.md for legal info.

#include "path.h"

#include "compiler.h"
#include "fslike/directory.h"
#include "fslike/native.h"
#include "fslike/python.h"
#include "misc.h"
#include "strings.h"
#include "../error/error.h"


namespace openage {
namespace util {


/**
 * Strip out ../ etc
 */
void path_normalizer(Path::parts_t &output, const Path::parts_t &input) {
	output.reserve(input.size());

	// normalize the path parts
	for (auto &part : input) {
		if (part == "." or part == "") {
			continue;
		}
		else if (part == "..") {
			if (likely(output.size() > 0)) {
				output.pop_back();
			}
		}
		else {
			output.push_back(part);
		}
	}
}


Path::Path() {}


Path::Path(py::Obj fsobj_in,
           const std::vector<std::string> &parts) {

	path_normalizer(this->parts, parts);

	// optimization: the fsobj from python may be convertible to
	//               a native c++ implementation.
	//               then calls don't need to be relayed any more.

	// test if fsobj is fslike.Directory
	if (fslike::pyx_fs_is_fslike_directory.call(fsobj_in.get_ref())) {
		this->fsobj = std::make_shared<fslike::Directory>(
			fsobj_in.getattr("path").bytes()
		);
	}
	else {
		// we can't create a c++-variant of the path,
		// so just wrap the python path.
		this->fsobj = std::make_shared<fslike::Python>(fsobj_in);
	}
}


Path::Path(const std::shared_ptr<fslike::FSLike> &fsobj,
           const parts_t &parts)
	:
	fsobj{fsobj} {

	path_normalizer(this->parts, parts);
}


bool Path::exists() const {
	return this->is_file() or this->is_dir();
}

bool Path::is_file() const {
	return this->fsobj->is_file(this->parts);
}

bool Path::is_dir() const {
	return this->fsobj->is_dir(this->parts);
}

bool Path::writable() const {
	return this->fsobj->writable(this->parts);
}

std::vector<Path::part_t> Path::list() {
	return this->fsobj->list(this->parts);
}

std::vector<Path> Path::iterdir() {
	std::vector<Path> ret;

	// return a new path object for each element in that directory
	for (auto &entry : this->fsobj->list(this->parts)) {
		ret.push_back(this->joinpath(entry));
	}

	return ret;
}

bool Path::mkdirs() {
	return this->fsobj->mkdirs(this->parts);
}

File Path::open(const std::string &mode) const {
	if (mode == "r") {
		return this->open_r();
	}
	else if (mode == "w") {
		return this->open_w();
	}
	else {
		throw Error{ERR << "unsupported open mode: " << mode};
	}
}


File Path::open_r() const {
	return this->fsobj->open_r(this->parts);
}


File Path::open_w() const {
	return this->fsobj->open_w(this->parts);
}


std::string Path::get_native_path() const {
	return this->fsobj->get_native_path(this->parts);
}


std::string Path::resolve_native_path(const std::string &mode) const {
	if (mode == "r") {
		return this->resolve_native_path_r();
	}
	else if (mode == "w") {
		return this->resolve_native_path_w();
	}
	else {
		throw Error{ERR << "unsupported resolve mode: " << mode};
	}
}


std::string Path::resolve_native_path_r() const {
	auto resolved_path = this->fsobj->resolve_r(this->parts);

	if (resolved_path.first) {
		return resolved_path.second.get_native_path();
	}
	else {
		throw Error{ERR << "failed to locate readable path: " << *this};
	}
}


std::string Path::resolve_native_path_w() const {
	auto resolved_path = this->fsobj->resolve_w(this->parts);

	if (resolved_path.first) {
		return resolved_path.second.get_native_path();
	}
	else {
		throw Error{ERR << "failed to locate writable path: " << *this};
	}
}


bool Path::rename(const Path &target_path) {
	// check if both fsobj pointers are the same
	// maybe a real equality check is better?
	if (this->fsobj != target_path.fsobj) {
		throw Error{
			ERR << "can't rename across two different filesystem like objects"
		};
	}
	return this->fsobj->rename(this->parts, target_path.parts);
}

bool Path::rmdir() {
	return this->fsobj->rmdir(this->parts);
}

bool Path::touch() {
	return this->fsobj->touch(this->parts);
}

bool Path::unlink() {
	return this->fsobj->unlink(this->parts);
}

void Path::removerecursive() {
	if (this->is_dir()) {
		for (auto &path : this->iterdir()) {
			path.removerecursive();
		}
		this->rmdir();
	}
	else {
		this->unlink();
	}
}


int Path::get_mtime() const {
	return this->fsobj->get_mtime(this->parts);
}

uint64_t Path::get_filesize() const {
	return this->fsobj->get_filesize(this->parts);
}

// int Path::watch();
// void Path::poll_fs_watches();

Path Path::get_parent() const {
	return this->joinpath("..");
}


const std::string &Path::get_name() const {
	if (this->parts.size() > 0) {
		return this->parts.back();
	}
	else {
		return util::empty_string;
	}
}

std::string Path::get_suffix() const {
	auto &name = this->get_name();
	size_t pos = name.rfind(".");
	if (pos == std::string::npos) {
		return "";
	}

	return name.substr(pos);
}

std::vector<std::string> Path::get_suffixes() const {
	std::string name = this->get_name();
	if (name.size() > 0 and name[0] == '.') {
		name = name.substr(1);
	}

	std::vector<std::string> ret;
	// ret = ['.' + suffix for suffix in name.split('.')[1:]]
	return ret;
}

std::string Path::get_stem() const {
	const std::string &name = this->get_name();
	size_t pos = name.rfind('.');
	if (pos == std::string::npos) {
		return name;
	}

	return name.substr(0, pos);
}


Path Path::joinpath(const parts_t &subpaths) const {
	parts_t new_parts = this->parts;
	for (auto &part : subpaths) {
		if (likely(part.size() > 0)) {
			new_parts.push_back(part);
		}
	}
	return Path{this->fsobj, new_parts};
}

Path Path::joinpath(const part_t &subpath) const {
	return this->joinpath(util::split(subpath, '/'));
}

Path Path::operator [](const parts_t &subpaths) const {
	return this->joinpath(subpaths);
}

Path Path::operator [](const part_t &subpath) const {
	return this->joinpath(subpath);
}

Path Path::operator /(const part_t &subpath) const {
	return this->joinpath(subpath);
}

Path Path::with_name(const part_t &name) const {
	return this->get_parent().joinpath(name);
}

Path Path::with_suffix(const part_t &suffix) const {
	return this->with_name(this->get_stem() + suffix);
}

bool Path::operator ==(const Path &other) const {
	return (
		this->fsobj == other.fsobj and
		this->parts == other.parts
	);
}

bool Path::operator !=(const Path &other) const {
	return not (*this == other);
}


fslike::FSLike *Path::get_fsobj() const {
	return this->fsobj.get();
}


const Path::parts_t &Path::get_parts() const {
	return this->parts;
}


std::ostream &operator <<(std::ostream &stream, const Path &path) {
	stream << "Path(";
	path.fsobj->repr(stream);
	stream << ":";

	for (auto &part : path.parts) {
		stream << "/" << part;
	}

	stream << ")";

	return stream;
}


std::string filename(const std::string &fullpath) {
	size_t rsep_pos = fullpath.rfind(fslike::PATHSEP);
	if (rsep_pos == std::string::npos) {
		return "";
	} else {
		return fullpath.substr(rsep_pos + 1);
	}
}


std::string dirname(const std::string &fullpath) {
	size_t rsep_pos = fullpath.rfind(fslike::PATHSEP);
	if (rsep_pos == std::string::npos) {
		return "";
	} else {
		return fullpath.substr(0, rsep_pos);
	}
}


}} // openage::util
